google/cloud/bigtable/instance_admin_test.cc

```diff
@@ -1427,3 +1427,244 @@ TEST_F(InstanceAdminTest, UpdateClusterPollRecoverableFailures) {
+      state: READY
+      serve_nodes: 0
+      default_storage_type: SSD
+  )";
+  btproto::Cluster expected;
+  ASSERT_TRUE(
+      google::protobuf::TextFormat::ParseFromString(expected_text, &expected));
+
+  btproto::Cluster expected_copy;
+  expected_copy.CopyFrom(expected);
+  bigtable::ClusterConfig cluster_config(std::move(expected));
+
+  EXPECT_CALL(*client_, UpdateCluster(_, _, _))
+      .WillOnce(Invoke([&expected_copy](
+                           grpc::ClientContext*,
+                           btproto::Cluster const& request,
+                           google::longrunning::Operation* response) {
+        auto const cluster_name =
+            "projects/my-project/instances/test-instance/clusters/test-cluster";
+        EXPECT_EQ(cluster_name, request.name());
+        response->set_done(true);
+        response->set_name("operation-name");
+        auto any = bigtable::internal::make_unique<google::protobuf::Any>();
+        any->PackFrom(expected_copy);
+        response->set_allocated_response(any.release());
+        return grpc::Status::OK;
+      }));
+
+  EXPECT_CALL(*client_, GetOperation(_, _, _)).Times(0);
+
+  auto future = tested.UpdateCluster(std::move(cluster_config));
+  auto actual = future.get();
+
+  std::string delta;
+  google::protobuf::util::MessageDifferencer differencer;
+  differencer.ReportDifferencesToString(&delta);
+  EXPECT_TRUE(differencer.Compare(expected_copy, actual)) << delta;
+}
+
+/// @test Failures while polling in `bigtable::InstanceAdmin::UpdateAppProfile`.
+TEST_F(InstanceAdminTest, UpdateAppProfilePollRecoverableFailures) {
+  using ::testing::_;
+  using ::testing::Invoke;
+
+  bigtable::InstanceAdmin tested(client_);
+  EXPECT_CALL(*client_, UpdateCluster(_, _, _))
+      .WillOnce(Invoke([](grpc::ClientContext*, btproto::Cluster const& request,
+                          google::longrunning::Operation* response) {
+        auto const cluster_name =
+            "projects/my-project/instances/test-instance/clusters/test-cluster";
+        EXPECT_EQ(cluster_name, request.name());
+        return grpc::Status::OK;
+      }));
+
+  std::string expected_text = R"(
+      name: 'projects/my-project/instances/test-instance/clusters/test-cluster'
+      location: 'Location1'
+      state: READY
+      serve_nodes: 0
+      default_storage_type: SSD
+  )";
+  btproto::Cluster expected;
+  ASSERT_TRUE(
+      google::protobuf::TextFormat::ParseFromString(expected_text, &expected));
+
+  btproto::Cluster expected_copy;
+  expected_copy.CopyFrom(expected);
+
+  bigtable::ClusterConfig cluster_config(std::move(expected));
+
+  EXPECT_CALL(*client_, GetOperation(_, _, _))
+      .WillOnce(Invoke([](grpc::ClientContext*,
+                          google::longrunning::GetOperationRequest const&,
+                          google::longrunning::Operation*) {
+        return grpc::Status(grpc::StatusCode::UNAVAILABLE, "try-again");
+      }))
+      .WillOnce(Invoke([](grpc::ClientContext*,
```

Comment:
This is doing the identical step twice... does this mean it should be

```cpp
.Will(Invoke(...)).Times(2)
```

instead of a copy-pasted block with `.WillOnce(...)`?

Reply:
Sorry, I got the methods wrong, it should be:

```cpp
.Times(2).WillRepeatedly(...)
```

See [the docs](https://github.com/google/googletest/blob/master/googlemock/docs/CheatSheet.md#using-mocks-in-tests), here's their code sample:

```cpp
  EXPECT_CALL(foo, Describe(5))                   // #4
      .Times(3)
      .WillRepeatedly(Return("Category 5"));
```

Reply:
Right, but you do this:

```C++
  EXPECT_CALL(*client_, UpdateAppProfile(_, _, _))
      .Times(3)
      .WillRepeatedly(
          Return(grpc::Status(grpc::StatusCode::UNAVAILABLE, "try-again")))
      .WillOnce(Invoke(...));
```

you get this:

```
 Failure.WillOnce() cannot appear after .WillRepeatedly() or .RetiresOnSaturation().
```

and this does not work either:

```C++
  EXPECT_CALL(*client_, UpdateAppProfile(_, _, _))
      .Times(3)
      .WillRepeatedly(
          Return(grpc::Status(grpc::StatusCode::UNAVAILABLE, "try-again")))
      .Times(1)
      .WillRepeatedly(Invoke(
```

That fails with:

```
 Failure.Times() cannot appear after .InSequence(), .WillOnce(), .WillRepeatedly(), or .RetiresOnSaturation().
```

I also tried making two calls to `EXPECT_THAT()` that fails because the second overrides the first I believe.  I guess we could use `InSequence()` but then things are getting far more complicated than repeating the `Return...` stuff (I think).

Reply:
OK, my bad, never mind then. Looks like you already have the simplest solution, given the constraints.